docs/design_proposals/log-selectors.md

```diff
@@ -0,0 +1,102 @@
+# Title
+
+* Author(s): Corneius Weig (@corneliusweig)
+* Design Shepherd: Priya Wadhwa (@priyawadhwa)
+* Date: 21/04/2019
+* Status: [Reviewed/Cancelled/Under implementation/Complete]
+
+## Background
+
+Skaffold offers the possibility to watch pod logs for the `dev` and `run` subcommands.
+So far, the pods to watch are determined by looking for known artifact images in the containers.
+This is restricting for two reasons:
+1. It does not allow to add additional pods to log aggregator which are not deployed by Skaffold (#666).
+2. It does not allow to exclude uninteresting pods deployed by Skaffold. For example, when working with istio, all the irrelevant istio logs spam the log (#588) 
+
+## Design
+I suggest to change the selection of pods for the log aggregator based on labels on this pod.
+
+For that, we need to change the definition of the `SkaffoldRunner` as follows:
+```go
+type SkaffoldRunner struct {
+	build.Builder
+	deploy.Deployer
+	test.Tester
+	tag.Tagger
+	sync.Syncer
+	watch.Watcher
+
+	cache             *cache.Cache
+	opts              *config.SkaffoldOptions
+	labellers         []deploy.Labeller
+	builds            []build.Artifact
+	hasBuilt          bool
+	hasDeployed       bool
+	
+	// <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
+	// OLD
+	// imageList         *kubernetes.ImageList
+	// NEW
+	podSelector       kubernetes.PodSelector
+	// <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
+	
+	RPCServerShutdown func() error
+}
+```
+This change does not affect the public API.
+
+To support the new use-cases of altering the pod selector, I propose a new top-level config section in the pipeline config:
+```yaml
+log:
+  # a list of deployment.spec.selector items, rules are OR'ed
+  - matchExpressions: # in one item, rules are AND'ed
```

Comment:
Just to clarify -- I would either specify `matchExpressions` or `matchLabels`? and will `matchExpressions` look for matches with labels on pods, or something else?

Reply:
@priyawadhwa Yes, `matchExpressions` and `matchLabels` should be marked with `yamltags:oneOf=selector` or so. I'll add that.
Both `matchExpression` and `matchLabels` should have the same semantics as for deployments, so they both refer to labels -- only `matchExpression` will allow more logical operations than just `matchLabels`.

Reply:
On second thoughts, after reading `kubectl explain deployment.spec.selector` carefully again, I don't think we should place the `oneOf` constraint here. `matchLabels` is simply a shorthand for `matchExpressions` and can be trivially AND'ed with a `matchExpressions` object.

Reply:
Got it. The nice thing about `matchLabels` is that it's simpler, but `matchExpressions` is essentially a superset of `matchLabels`. I'm wondering if it would be cleaner to only have `matchExpressions`.

We could have something like 
```yaml
- key: tail
  values: ['true']
```
and then default the `operator` to `In` so that it would be a bit simpler like `matchLabels`. WDYT?

Reply:
Got it. The nice thing about `matchLabels` is that it's simpler, but `matchExpressions` is essentially a superset of `matchLabels`. I'm wondering if it would be cleaner to only have `matchExpressions`.

We could have something like 
```yaml
- key: tail
  values: ['true']
```
and then default the `operator` to `In` so that it would be a bit simpler like `matchLabels`. WDYT?

Reply:
Not quite convinced.
If Skaffold wants to look like k8s selectors, we should strive to be as compliant as possible. Using similar names with deviating semantics (default operator) creates confusion. The great advantage of using the standard syntax and semantics for the selector is the recognizability with existing k8s syntax.
Bottom line: if you really think we should deviate from the standard k8s config, then the new config should also look different. But why should it? I think k8s has found a sweet spot of configurability (`matchExpressions`) and simplicity (`matchLabels`).

Reply:
Fair point, I don't feel super strongly about it, so let's keep it the way you have it!